dev/tests/integration/testsuite/Magento/GraphQl/Controller/CorsGraphQlTest.php

```diff
@@ -0,0 +1,132 @@
+    {
+        $this->addOriginAndSendGraphQlRequest("https://www.example.com");
+        $response = $this->getResponse();
+        $this->assertNotFalse($response->getHeader('Access-Control-Allow-Origin'));
+        $this->assertNotFalse($response->getHeader('Access-Control-Allow-Headers'));
+        $this->assertNotFalse($response->getHeader('Access-Control-Allow-Methods'));
+        $this->assertNotFalse($response->getHeader('Access-Control-Max-Age'));
+    }
+
+    /**
+     * @magentoConfigFixture default/web/graphql/cors_allowed_origins https://www.example.com
+     * @magentoConfigFixture default/web/graphql/cors_allowed_headers Content-Type
+     * @magentoConfigFixture default/web/graphql/cors_allowed_methods GET,POST,OPTIONS
+     * @magentoConfigFixture default/web/graphql/cors_max_age 86400
+     * @magentoConfigFixture default/web/graphql/cors_allow_credentials 1
+     */
+    public function testNormalRequestDoesNotContainsCorsHeaders()
+    {
+        $httpHeaders = new Headers();
+        $httpHeaders->addHeaderLine('Origin: https://www.example.com');
+        $this->dispatch('/');
+
+        $response = $this->getResponse();
+        $this->assertFalse($response->getHeader('Access-Control-Allow-Origin'));
+        $this->assertFalse($response->getHeader('Access-Control-Allow-Headers'));
+        $this->assertFalse($response->getHeader('Access-Control-Allow-Methods'));
+        $this->assertFalse($response->getHeader('Access-Control-Max-Age'));
+    }
+
+    /**
+     * @magentoConfigFixture default/web/graphql/cors_allowed_origins https://www.example.com
+     * @magentoConfigFixture default/web/graphql/cors_allowed_headers Content-Type
+     * @magentoConfigFixture default/web/graphql/cors_allowed_methods GET,POST,OPTIONS
+     * @magentoConfigFixture default/web/graphql/cors_max_age 86400
+     * @magentoConfigFixture default/web/graphql/cors_allow_credentials 1
+     */
+    public function testCorsNotAddedIfOriginIsNotAllowed()
+    {
+        $this->addOriginAndSendGraphQlRequest("https://www.test.com");
+        $response = $this->getResponse();
+        $this->assertFalse($response->getHeader('Access-Control-Allow-Origin'));
+        $this->assertFalse($response->getHeader('Access-Control-Allow-Headers'));
+        $this->assertFalse($response->getHeader('Access-Control-Allow-Methods'));
+        $this->assertFalse($response->getHeader('Access-Control-Max-Age'));
+    }
+
+    public function testCorsRequestFailsIfCorsConfigurationIsNotProvided()
+    {
+        $this->addOriginAndSendGraphQlRequest("https://www.example.com");
+        $response = $this->getResponse();
+        $this->assertFalse($response->getHeader('Access-Control-Allow-Origin'));
+        $this->assertFalse($response->getHeader('Access-Control-Allow-Headers'));
+        $this->assertFalse($response->getHeader('Access-Control-Allow-Methods'));
+        $this->assertFalse($response->getHeader('Access-Control-Max-Age'));
+    }
```

Comment:
It'd be great if you could add an additional use case by setting the http method "OPTIONS" to the request since CORS allow this method. 
Thanks

Reply:
Added OPTIONS test case with error validation.